{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comment (2023.01.18.)\n",
    "\n",
    "# .py & .ipynb 차이점\n",
    "# - MQpy/old_notebooks/pg_03_3_processing.ipynb 에서 시작.\n",
    "# - 복붙 x. 타이핑 해가며 기억 깨우기...!\n",
    "# - directory 달라져서 변경해야함. raw_files 폴더 위치 파일 사용하기.\n",
    "# .py 일 경우, 파이썬 패키지 설치하기 위해서 환경변수 설정해야. 파이썬 설치해야.\n",
    "# .ipynb 일 경우, 기본 탑재.\n",
    "\n",
    "\n",
    "# Scope\n",
    "\n",
    "# - proteinGroups.txt 로드 후 다음의 column data를 필터함.\n",
    "# : contaminants, reverse, only identified site, uniquepeptide = 1 entries\n",
    "# - protein names, Best MS/MS 항목에서 세미콜론(;) 구분된 데이터 요소를 split.\n",
    "# - filter가 된 것을 *_base.xlsx 로 저장함.\n",
    "\n",
    "\n",
    "# Uniprot API\n",
    "\n",
    "# - IdMapping: https://uniprot.org/help/id_mapping 참조\n",
    "\n",
    "\n",
    "# proteinGroups.txt 의 Dtype\n",
    "\n",
    "# - object(delimeter), int64, float64\n",
    "\n",
    "\n",
    "# vscode 터미널에서 pip 안될때\n",
    "\n",
    "# 원인: python 옳게 설치했더라도 윈도우 클래스에서 pip 위치를 알지 못하기에 발생.\n",
    "# python3 부터는 pip 내장이므로 python.exe 디렉터리 연결하면 됨.\n",
    "# C:\\Users\\Simon\\AppData\\Local\\Programs\\Python\\Python39\\Scipts 를 다음에 추가하면됨.\n",
    "# (고급 시스템 설정 보기) - (환경변수 탭) - (사용자에 대한 사용자 변수 및 시스템 변수의 path에 추가)\n",
    "\n",
    "\n",
    "# 아나콘다 업데이트\n",
    "# conda update -n base conda <-- base 업데이트\n",
    "# conda update --all <-- 패키지 업데이트\n",
    "# python -m pip install --upgrade pip <-- pip 패키지 업데이트\n",
    "# conda --version <-- version 확인\n",
    "\n",
    "\n",
    "# 다른 설정들\n",
    "# 셀 라인 표시 shift + L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "message! >>> 11 (0.92%) entries were dropped, [Potential contaminant] column removed.\n",
      "message! >>> 12 (1.01%) entries were dropped, [Reverse] column removed.\n",
      "message! >>> 25 (2.13%) entries were dropped, [Only identified by site] column removed.\n",
      "message! >>> 385 (33.60%) entries were dropped. [Razor + unique peptides = 1]\n",
      "message! >>> 761 entries left.\n",
      "message! >>> [Protein IDs] elements were splitted\n",
      "message! >>> [Best MS/MS] elements were splitted\n",
      "<bound method NDFrame.head of      Protein IDs Majority protein IDs Peptide counts (all)  \\\n",
      "4         A0FGR8               A0FGR8                    2   \n",
      "5         A6NHR9               A6NHR9                    2   \n",
      "7         Q99613        Q99613;B5ME19                  4;4   \n",
      "18        E9PAV3               E9PAV3                  5;1   \n",
      "20        O00148               O00148                    8   \n",
      "...          ...                  ...                  ...   \n",
      "1175      Q9Y5S9               Q9Y5S9                    4   \n",
      "1176      Q9Y617               Q9Y617                   13   \n",
      "1178      Q9Y678               Q9Y678                  3;1   \n",
      "1179      Q9Y6N5               Q9Y6N5                    5   \n",
      "1181      Q9Y6Y8               Q9Y6Y8                    2   \n",
      "\n",
      "     Peptide counts (razor+unique) Peptide counts (unique)  \\\n",
      "4                                2                       2   \n",
      "5                                2                       2   \n",
      "7                              4;4                     4;4   \n",
      "18                             5;1                     5;1   \n",
      "20                               2                       2   \n",
      "...                            ...                     ...   \n",
      "1175                             4                       4   \n",
      "1176                            13                      13   \n",
      "1178                           3;1                     3;1   \n",
      "1179                             5                       5   \n",
      "1181                             2                       2   \n",
      "\n",
      "                                          Protein names    Gene names  \\\n",
      "4                              Extended synaptotagmin-2         ESYT2   \n",
      "5     Structural maintenance of chromosomes flexible...        SMCHD1   \n",
      "7     Eukaryotic translation initiation factor 3 sub...  EIF3C;EIF3CL   \n",
      "18    Nascent polypeptide-associated complex subunit...          NACA   \n",
      "20                    ATP-dependent RNA helicase DDX39A        DDX39A   \n",
      "...                                                 ...           ...   \n",
      "1175                             RNA-binding protein 8A         RBM8A   \n",
      "1176                     Phosphoserine aminotransferase         PSAT1   \n",
      "1178                           Coatomer subunit gamma-1         COPG1   \n",
      "1179      Sulfide:quinone oxidoreductase, mitochondrial         SQRDL   \n",
      "1181                          SEC23-interacting protein       SEC23IP   \n",
      "\n",
      "                                          Fasta headers  Number of proteins  \\\n",
      "4     sp|A0FGR8|ESYT2_HUMAN Extended synaptotagmin-2...                   1   \n",
      "5     sp|A6NHR9|SMHD1_HUMAN Structural maintenance o...                   1   \n",
      "7     sp|Q99613|EIF3C_HUMAN Eukaryotic translation i...                   2   \n",
      "18    sp|E9PAV3|NACAM_HUMAN Nascent polypeptide-asso...                   2   \n",
      "20    sp|O00148|DX39A_HUMAN ATP-dependent RNA helica...                   1   \n",
      "...                                                 ...                 ...   \n",
      "1175  sp|Q9Y5S9|RBM8A_HUMAN RNA-binding protein 8A O...                   1   \n",
      "1176  sp|Q9Y617|SERC_HUMAN Phosphoserine aminotransf...                   1   \n",
      "1178  sp|Q9Y678|COPG1_HUMAN Coatomer subunit gamma-1...                   2   \n",
      "1179  sp|Q9Y6N5|SQOR_HUMAN Sulfide:quinone oxidoredu...                   1   \n",
      "1181  sp|Q9Y6Y8|S23IP_HUMAN SEC23-interacting protei...                   1   \n",
      "\n",
      "      Peptides  ...  MS/MS count    id  \\\n",
      "4            2  ...            3     4   \n",
      "5            2  ...            3     5   \n",
      "7            4  ...            5     7   \n",
      "18           5  ...           14    18   \n",
      "20           8  ...            2    20   \n",
      "...        ...  ...          ...   ...   \n",
      "1175         4  ...            4  1175   \n",
      "1176        13  ...           24  1176   \n",
      "1178         3  ...            3  1178   \n",
      "1179         5  ...            6  1179   \n",
      "1181         2  ...            3  1181   \n",
      "\n",
      "                                            Peptide IDs  \\\n",
      "4                                             2721;4259   \n",
      "5                                             2770;5114   \n",
      "7                                    985;1470;1767;2166   \n",
      "18                              584;2064;3346;3427;3968   \n",
      "20                 456;546;946;1641;1758;1898;2203;2976   \n",
      "...                                                 ...   \n",
      "1175                                   73;811;1859;3342   \n",
      "1176  204;326;425;524;1242;1276;1462;2143;2865;3494;...   \n",
      "1178                                     2208;3886;3921   \n",
      "1179                           1899;2219;3567;3604;4909   \n",
      "1181                                          1600;1894   \n",
      "\n",
      "                                       Peptide is razor  \\\n",
      "4                                             True;True   \n",
      "5                                             True;True   \n",
      "7                                   True;True;True;True   \n",
      "18                             True;True;True;True;True   \n",
      "20        True;False;False;False;False;True;False;False   \n",
      "...                                                 ...   \n",
      "1175                                True;True;True;True   \n",
      "1176  True;True;True;True;True;True;True;True;True;T...   \n",
      "1178                                     True;True;True   \n",
      "1179                           True;True;True;True;True   \n",
      "1181                                          True;True   \n",
      "\n",
      "                                       Mod. peptide IDs  \\\n",
      "4                                             2758;4311   \n",
      "5                                             2807;5175   \n",
      "7                                   1003;1492;1794;2196   \n",
      "18                              596;2092;3393;3475;4019   \n",
      "20                 465;558;964;1667;1785;1925;2234;3016   \n",
      "...                                                 ...   \n",
      "1175                                   74;827;1886;3389   \n",
      "1176  208;332;434;535;1263;1297;1484;2172;2905;3542;...   \n",
      "1178                                     2239;3936;3971   \n",
      "1179                           1926;2250;3615;3652;4967   \n",
      "1181                                          1626;1921   \n",
      "\n",
      "                                           Evidence IDs  \\\n",
      "4                                             3143;4944   \n",
      "5                                             3198;5925   \n",
      "7                                   1140;1674;2037;2502   \n",
      "18                    681;2381;2382;3857;3858;3953;4563   \n",
      "20                535;639;1095;1895;2027;2188;2545;3428   \n",
      "...                                                 ...   \n",
      "1175                                   83;937;2145;3851   \n",
      "1176  242;383;498;614;1431;1466;1665;2473;3308;4029;...   \n",
      "1178                                     2550;4466;4510   \n",
      "1179                           2189;2562;4106;4149;5689   \n",
      "1181                                          1838;2184   \n",
      "\n",
      "                                              MS/MS IDs Best MS/MS  \\\n",
      "4                                        4991;7829;7830       4991   \n",
      "5                                        5064;5065;9468       5065   \n",
      "7                              1824;1825;2643;3221;4009       1825   \n",
      "18    1107;1108;3807;3808;3809;6060;6061;6062;6063;6...       1108   \n",
      "20          873;1048;1760;3009;3206;3486;4075;4076;5407        873   \n",
      "...                                                 ...        ...   \n",
      "1175                            124;1519;3413;3414;6052        124   \n",
      "1176  383;384;626;627;806;1006;1007;2270;2324;2325;2...        383   \n",
      "1178                                     4082;7001;7080       4082   \n",
      "1179                 3487;4105;4106;6469;6536;6537;9066       3487   \n",
      "1181                                     2909;2910;3481       2910   \n",
      "\n",
      "      Oxidation (M) site IDs  Oxidation (M) site positions  \n",
      "4                        NaN                           NaN  \n",
      "5                        NaN                           NaN  \n",
      "7                        NaN                           NaN  \n",
      "18                       NaN                           NaN  \n",
      "20                       NaN                           NaN  \n",
      "...                      ...                           ...  \n",
      "1175                     NaN                           NaN  \n",
      "1176                     NaN                           NaN  \n",
      "1178                     NaN                           NaN  \n",
      "1179                     NaN                           NaN  \n",
      "1181                     NaN                           NaN  \n",
      "\n",
      "[761 rows x 31 columns]>\n"
     ]
    }
   ],
   "source": [
    "# Neccessary packages\n",
    "import pandas as pd\n",
    "from time import localtime, strftime\n",
    "\n",
    "\n",
    "\n",
    "# Base_filter를 사용할 txt 베이스 raw 파일을 로드합니다.\n",
    "global txtpath\n",
    "txtpath = '..\\\\raw_files\\\\TestSample\\\\LTQ_QC_DU145\\\\proteinGroups.txt'\n",
    "\n",
    "\n",
    "'''\n",
    "class drop_column:\n",
    "    def __init__(self, c_name=None, identifier=None):\n",
    "        self.c_name = c_name\n",
    "        self.identifier = identifier\n",
    "\n",
    "    # 이 함수는 unique peptide 컬럼도 드랍하므로 사용 불가.\n",
    "    def IsDrop(self):\n",
    "        tmp1 = len(df)\n",
    "        tmp2 = len(df[df[self.c_name] == self.identifier])\n",
    "\n",
    "        df.drop(df[df[self.c_name] == self.identifier].index, inplace = True)\n",
    "        df.drop(columns=[self.c_name], inplace=True)\n",
    "        # 단, unique peptide에서는 필터후 컬럼 드랍하면 안돼.\n",
    "\n",
    "        ratio = (100*tmp2)/(tmp1)\n",
    "        return print('message! >>> '+str(tmp2)+' ['+self.c_name+'] (%.2f%%) of entries were dropped' %ratio)'''\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "# **kwargs 인수로 받는 함수... 클래스에 들어가면 init 잘 해야.\n",
    "def isDrop(**kwargs):\n",
    "    # **kwargs: keyword argument 줄임말. 인수를 딕셔너리로 받음.\n",
    "    for key, value in kwargs.items():\n",
    "        tmp1 = len(df)\n",
    "        tmp2 = len(df[df[key] == value])\n",
    "        ratio = (100*tmp2)/tmp1\n",
    "        \n",
    "        # key에 해당하는 value를 가진 entry 드랍.\n",
    "        df.drop(df[df[key] == value].index, inplace=True)\n",
    "        \n",
    "        # value가 숫자인 경우: column 드랍하지 않음.\n",
    "        # e.g., Razor + unique peptides의 경우, value = 1 인 entry 드랍 하지만, 나머지 entry는 남아야하므로.\n",
    "        if isinstance(value, str):\n",
    "            df.drop(columns=[key], inplace=True)\n",
    "            print('message! >>> '+str(tmp2)+' (%.2f%%) entries were dropped, [' %ratio +key+'] column removed.')\n",
    "        else:\n",
    "            print('message! >>> '+str(tmp2)+' (%.2f%%) entries were dropped. ['%ratio +key+' = '+str(value)+']')\n",
    "            # print('message! >>> '+str(tmp2)+' (%.2f%%) entries were dropped, ['+key+'] column removed' %ratio) <- 이 구문은 동작안함. %~~ 이게 string으로 나눈 같은 구역에 있어햐 함.\n",
    "\n",
    "    return print('message! >>> '+str(len(df))+' entries left.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Split할 column 이름을 tuple (c1, c2)로 주고, delimeter의 기본값은 세미콜론(;)으로 되어있다.\n",
    "def split_items(*args, delimeter=';'):\n",
    "    for arg in args:\n",
    "        tmp_series = pd.Series(df[arg])\n",
    "        for ele in tmp_series:\n",
    "            tmp = ele.split(delimeter)[0]\n",
    "            tmp_series.replace(ele, tmp, inplace=True)\n",
    "        print('message! >>> ['+arg+'] elements were splitted')\n",
    "    return None\n",
    "    # return df.head\n",
    "\n",
    "\n",
    "\n",
    "def execute(drop_rule, split_cnames):\n",
    "    # read txt file.\n",
    "    isDrop(**drop_rule)\n",
    "    split_items(*split_cnames)\n",
    "    # return None\n",
    "    \n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    df = pd.read_table(filepath_or_buffer=txtpath, index_col=False)\n",
    "    base_filter = {'Potential contaminant':'+', 'Reverse':'+', 'Only identified by site':'+', 'Razor + unique peptides':1}\n",
    "    split_cnames = ('Protein IDs', 'Best MS/MS')\n",
    "    execute(base_filter, split_cnames)\n",
    "    # split_items(split_cnames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# requests module 사용하여 web request-repond 프로세스를 수행합니다.\n",
    "# uniprot accession 형식의 Protein IDs 를 Uniprot ID Mapping에 요청-응답 수행.\n",
    "# 데이터프레임의 protein/gene name 및 sequence 등 항목을 수정해야 합니다.\n",
    "\n",
    "# 참고페이지\n",
    "# Uniprot API Help\n",
    "\n",
    "import uniprot_requests as unireq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Simon\\AppData\\Local\\Temp\\ipykernel_20088\\2141442334.py:3: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.\n",
      "  idSeries = PRids[:6]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4     A0FGR8\n",
       "5     A6NHR9\n",
       "7     Q99613\n",
       "18    E9PAV3\n",
       "20    O00148\n",
       "21    O00151\n",
       "Name: Protein IDs, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 판다스로 데이터프레임의 컬럼 데이터 뽑아냅니다.\n",
    "# 테스트를 위하여 데이터의 일부를 새로운 객체 idSeries에 할당합니다.\n",
    "PRids = pd.Series(df['Protein IDs'])\n",
    "\n",
    "idSeries = PRids[:6]\n",
    "\n",
    "idSeries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrying in 3s\n",
      "Retrying in 3s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'https://rest.uniprot.org/idmapping/uniprotkb/results/ebd732214f1fa3b46504d64ad8382cb1b37a2cf0'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 테스트 데이터가 담긴 객체를 uniprot_request.py 모듈의 execute 메소드 호출 및 링크 받아옵니다.\n",
    "link = unireq.execute(idSeries)\n",
    "link"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0FGR8</td>\n",
       "      <td>A0FGR8</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ESYT2_HUMAN</td>\n",
       "      <td>Extended synaptotagmin-2 (E-Syt2) (Chr2Syt)</td>\n",
       "      <td>ESYT2 FAM62B KIAA1228</td>\n",
       "      <td>921</td>\n",
       "      <td>MTANRDAALSSHRHPGCAQRPRTPTFASSSQRRSAFGFDDGNFPGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A6NHR9</td>\n",
       "      <td>A6NHR9</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>SMHD1_HUMAN</td>\n",
       "      <td>Structural maintenance of chromosomes flexible...</td>\n",
       "      <td>SMCHD1 KIAA0650</td>\n",
       "      <td>2005</td>\n",
       "      <td>MAAADGGGPGGASVGTEEDGGGVGHRTVYLFDRREKESELGDRPLQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q99613</td>\n",
       "      <td>Q99613</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>EIF3C_HUMAN</td>\n",
       "      <td>Eukaryotic translation initiation factor 3 sub...</td>\n",
       "      <td>EIF3C EIF3S8</td>\n",
       "      <td>913</td>\n",
       "      <td>MSRFFTTGSDSESESSLSGEELVTKPVGGNYGKQPLLLSEDEEDTK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E9PAV3</td>\n",
       "      <td>E9PAV3</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NACAM_HUMAN</td>\n",
       "      <td>Nascent polypeptide-associated complex subunit...</td>\n",
       "      <td>NACA</td>\n",
       "      <td>2078</td>\n",
       "      <td>MPGEATETVPATEQELPQPQAETAVLPMSSALSVTAALGQPGPTLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O00148</td>\n",
       "      <td>O00148</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DX39A_HUMAN</td>\n",
       "      <td>ATP-dependent RNA helicase DDX39A (EC 3.6.4.13...</td>\n",
       "      <td>DDX39A DDX39</td>\n",
       "      <td>427</td>\n",
       "      <td>MAEQDVENDLLDYDEEEEPQAPQESTPAPPKKDIKGSYVSIHSSGF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     From   Entry  Reviewed   Entry Name  \\\n",
       "0  A0FGR8  A0FGR8  reviewed  ESYT2_HUMAN   \n",
       "1  A6NHR9  A6NHR9  reviewed  SMHD1_HUMAN   \n",
       "2  Q99613  Q99613  reviewed  EIF3C_HUMAN   \n",
       "3  E9PAV3  E9PAV3  reviewed  NACAM_HUMAN   \n",
       "4  O00148  O00148  reviewed  DX39A_HUMAN   \n",
       "\n",
       "                                       Protein names             Gene Names  \\\n",
       "0        Extended synaptotagmin-2 (E-Syt2) (Chr2Syt)  ESYT2 FAM62B KIAA1228   \n",
       "1  Structural maintenance of chromosomes flexible...        SMCHD1 KIAA0650   \n",
       "2  Eukaryotic translation initiation factor 3 sub...           EIF3C EIF3S8   \n",
       "3  Nascent polypeptide-associated complex subunit...                   NACA   \n",
       "4  ATP-dependent RNA helicase DDX39A (EC 3.6.4.13...           DDX39A DDX39   \n",
       "\n",
       "  Length                                           Sequence  \n",
       "0    921  MTANRDAALSSHRHPGCAQRPRTPTFASSSQRRSAFGFDDGNFPGL...  \n",
       "1   2005  MAAADGGGPGGASVGTEEDGGGVGHRTVYLFDRREKESELGDRPLQ...  \n",
       "2    913  MSRFFTTGSDSESESSLSGEELVTKPVGGNYGKQPLLLSEDEEDTK...  \n",
       "3   2078  MPGEATETVPATEQELPQPQAETAVLPMSSALSVTAALGQPGPTLP...  \n",
       "4    427  MAEQDVENDLLDYDEEEEPQAPQESTPAPPKKDIKGSYVSIHSSGF...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# link 통해 나온 결과는 stream.. 말그대로 텍스트의 엄청난 스트림!!\n",
    "# 같은 모듈의 get_id_mapping_results_stream 함수로 이 텍스트를 디코딩해서 필요한 부분을 추출합니다.\n",
    "#\n",
    "# ?compressed= true/false 는 다운로드 할 때 상관 있어서 차이는 없는 것으로 보임.\n",
    "# 추출: %2C[식별자] (e.g., %2Crequest, %2Cprotein_name, %2Cgene_name 등등.) 의 규칙으로 추출해서\n",
    "# 결과: &format=[확장자] (e.g., ~~&format=tsv) 로 결과물 만듦. delimeter 사용해서 데이터프레임 만든다.\n",
    "\n",
    "import csv\n",
    "import pandas as pd\n",
    "\n",
    "def get_data_frame_from_tsv_results(tsv_results):\n",
    "    reader = csv.DictReader(tsv_results, delimiter=\"\\t\", quotechar='\"')\n",
    "    return pd.DataFrame(list(reader))\n",
    "\n",
    "tsv_rst = unireq.get_id_mapping_results_stream(str(link)+'?compressed=true&fields=accession%2Creviewed%2Cid%2Cprotein_name%2Cgene_names%2Clength%2Csequence&format=tsv')\n",
    "\n",
    "tmp = get_data_frame_from_tsv_results(tsv_rst)\n",
    "\n",
    "tmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uniprot respond 데이터로 Protein names, Gene Names 교체할 수 있다.\n",
    "# 엑셀작업시 순서대로 였으면 순서대로 해도 되지만\n",
    "# 혹시 순서가 달라졌을수도 있으므로 From 값을 기준으로 replace를 수행할 것.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Protein IDs, Best MS/MS 의 delimeter (;) 로 구분된 데이터를 나누어 첫 번째 항목을 저장합니다.\n",
    "# 해당 column 이름을 1차원 series로 저장하고, deilmeter를 기준으로 구분한 뒤,\n",
    "# 첫번째 값으로 replace 합니다.\n",
    "# split에 사용한 객체는 초기화 되어야 합니다.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# (1) Protein IDs\n",
    "prot = pd.Series(df['Protein IDs'])\n",
    "for ele in prot:\n",
    "    tmp = ele.split(';')[0]\n",
    "    prot.replace(ele, tmp, inplace=True)\n",
    "print('message! >>> [Protein IDs] splitted.')\n",
    "\n",
    "# (2) Best MS/MS\n",
    "bmsms = pd.Series(df['Best MS/MS'])\n",
    "for ele in bmsms:\n",
    "    tmp = ele.split(';')[0]\n",
    "    bmsms.replace(ele, tmp, inplace=True)\n",
    "print('message! >>> [Best MS/MS] splitted.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Base 파일 생성하기.\n",
    "# 1단계로, protein groups에 대해서만 생성하고 2단계로 여러 txt 파일도 가능해야 합니다.\n",
    "# tab/comma 로 구분되는 csv, utf-8 encoding, index=False\n",
    "# 여기까지하면 예전에 작업한 pg_03_3_processing은 완료되었다 볼 수 있습니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop: Positive value for (Potential contaminant), (Reverse), (Only identified by site)\n",
    "# Drop : 1 (Razor + unique peptides)\n",
    "\n",
    "# Drop: Positive value for (Potential contaminant), and drop the column\n",
    "len_tmp = len(df[df['Potential contaminant'] == '+'])\n",
    "df.drop(df[df['Potential contaminant'] == '+'].index, inplace = True)\n",
    "df.drop(columns=['Potential contaminant'], inplace=True)\n",
    "print('message! >>> '+str(len_tmp)+' Potential contaminant entries were dropped')\n",
    "\n",
    "\n",
    "# Drop: Positive value for (Reverse), and drop the column\n",
    "len_tmp = len(df[df['Reverse'] == '+'])\n",
    "df.drop(df[df['Reverse'] == '+'].index, inplace = True)\n",
    "df.drop(columns=['Reverse'], inplace=True)\n",
    "print('message! >>> '+str(len_tmp)+' [Reverse entries] were dropped')\n",
    "\n",
    "\n",
    "# Drop: Positive value for (Only identified by site), and drop the column\n",
    "len_tmp = len(df[df['Only identified by site'] == '+'])\n",
    "df.drop(df[df['Only identified by site'] == '+'].index, inplace = True)\n",
    "df.drop(columns=['Only identified by site'], inplace=True)\n",
    "print('message! >>> '+str(len_tmp)+' [Only identified by site] entries were dropped')\n",
    "\n",
    "\n",
    "# Drop: (Razor + unique peptides = 1), and drop the column\n",
    "len_tmp = len(df[df['Razor + unique peptides'] == 1])\n",
    "df.drop(df[df['Razor + unique peptides'] == 1].index, inplace = True)\n",
    "df.drop(columns=['Razor + unique peptides'], inplace=True)\n",
    "print('message! >>> '+str(len_tmp)+' [Razor + uniqe peptides = 1] (%.1f%%) entries were dropped' %(100*len_tmp/(len(df)+len_tmp)))\n",
    "\n",
    "\n",
    "# Monitor: # of remained entries\n",
    "print('message! >>> '+str(len(df))+' entries were remained')\n",
    "\n",
    "\n",
    "# reset index\n",
    "df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Below... To Do List\n",
    "\n",
    "# 4가지 항목 드랍하는 구문 funciton 정의하여 짧게 구현. (완료)\n",
    "\n",
    "# Split [Protin IDs], [Best MS/MS] columns by semi-colon(;) (function 정의)\n",
    "\n",
    "# ID Mapping at Uniprot website export search result\n",
    "# then replace [Protein names], [Gene names], [sequence] values.\n",
    "# 여기서 bs4? API? 뭘 사용할까 -> curl 구문 사용.\n",
    "# Uniprot에서 python 전용 idmapping 풀코드를 지원해줬음. (완료)\n",
    "\n",
    "# Generate ProteinGroups_base.xlsx\n",
    "# In this .xlsx file, few columns left."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "4c37b9e9e27a0f58f1f76daeb0795aad7529a8acf1100c521e915f00a9c40e89"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
