{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "954de9d2-d1a6-4e4a-8895-5926b4deb37e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "message! >>> 73 Potential contaminant elements are dropped\n",
      "message! >>> 12 reverse elements are dropped\n",
      "message! >>> 2 columns are dropped: [Reverse] [Potential contaminant]\n",
      "message! >>> 29 columns are dropped\n",
      "message! >>> ./outputs/peptides_v1-20220627_163305.txt added.\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 5591 entries, 0 to 5590\n",
      "Data columns (total 26 columns):\n",
      " #   Column                  Non-Null Count  Dtype  \n",
      "---  ------                  --------------  -----  \n",
      " 0   Sequence                5591 non-null   object \n",
      " 1   N-term cleavage window  5591 non-null   object \n",
      " 2   C-term cleavage window  5591 non-null   object \n",
      " 3   Amino acid before       5591 non-null   object \n",
      " 4   Amino acid after        5591 non-null   object \n",
      " 5   Missed cleavages        5591 non-null   int64  \n",
      " 6   Mass                    5591 non-null   float64\n",
      " 7   Proteins                5591 non-null   object \n",
      " 8   Leading razor protein   5591 non-null   object \n",
      " 9   Start position          5591 non-null   float64\n",
      " 10  End position            5591 non-null   float64\n",
      " 11  Gene names              5588 non-null   object \n",
      " 12  Protein names           5588 non-null   object \n",
      " 13  Unique (Groups)         5591 non-null   object \n",
      " 14  Unique (Proteins)       5591 non-null   object \n",
      " 15  Charges                 5591 non-null   object \n",
      " 16  PEP                     5591 non-null   float64\n",
      " 17  Score                   5591 non-null   float64\n",
      " 18  Intensity               5591 non-null   int64  \n",
      " 19  id                      5591 non-null   int64  \n",
      " 20  Protein group IDs       5591 non-null   object \n",
      " 21  Mod. peptide IDs        5591 non-null   object \n",
      " 22  Evidence IDs            5591 non-null   object \n",
      " 23  MS/MS IDs               5591 non-null   object \n",
      " 24  Best MS/MS              5591 non-null   int64  \n",
      " 25  MS/MS Count             5591 non-null   int64  \n",
      "dtypes: float64(5), int64(5), object(16)\n",
      "memory usage: 1.1+ MB\n"
     ]
    }
   ],
   "source": [
    "# peptides.txt 전처리 구문\n",
    "# v0.0: contam, reverse 필터링\n",
    "# v1.0: 선택 column제외 필터링\n",
    "\n",
    "import pandas as pd\n",
    "from time import localtime, strftime\n",
    "\n",
    "global txtpath\n",
    "txtpath = 'F:/HFX-2022-DATA-Searched/20220615_DU145_Global_100ng_60min_MQ2.1/combined/txt/peptides.txt'\n",
    "\n",
    "class op_address:\n",
    "    def __init__(self, vers):\n",
    "        self.vers = vers\n",
    "        \n",
    "    def pg_address(self):\n",
    "        address = './outputs/'+txtpath[26:65]+'_proteinGroups_v'+str(self.vers)+'.txt'\n",
    "        return address\n",
    "\n",
    "    def pep_address(self):\n",
    "        address = './outputs/'+txtpath[26:65]+'_peptides_v'+str(self.vers)+'.txt'\n",
    "        return address\n",
    "    \n",
    "    def pg_address2(self):\n",
    "        tm = localtime()\n",
    "        ntm = strftime('%Y%m%d_%H%M%S', tm)\n",
    "        address = './outputs/proteinGroups_v'+str(self.vers)+'-'+ntm+'.txt'\n",
    "        return address\n",
    "    \n",
    "    def pep_address2(self):\n",
    "        tm = localtime()\n",
    "        ntm = strftime('%Y%m%d_%H%M%S', tm)\n",
    "        address = './outputs/peptides_v'+str(self.vers)+'-'+ntm+'.txt'\n",
    "        return address\n",
    "\n",
    "\n",
    "\n",
    "# peptides.txt 로드.\n",
    "df = pd.read_table(filepath_or_buffer=txtpath)\n",
    "len0 = len(df)\n",
    "\n",
    "# contam, reverse, only site 필터링 후 column drop\n",
    "df.drop(df[df['Potential contaminant'] == '+'].index, inplace = True)\n",
    "len_contam = len0 - len(df)\n",
    "print('message! >>> '+str(len_contam)+' Potential contaminant elements are dropped')\n",
    "\n",
    "df.drop(df[df['Reverse'] == '+'].index, inplace = True)\n",
    "len_rev = len0-len_contam-len(df)\n",
    "print('message! >>> '+str(len_rev)+' reverse elements are dropped')\n",
    "      \n",
    "df.drop(columns=['Reverse', 'Potential contaminant'], inplace=True)\n",
    "print('message! >>> 2 columns are dropped: [Reverse] [Potential contaminant]')\n",
    "\n",
    "\n",
    "# version 0.0 : Contaminatant, Reverse elimination.\n",
    "# df.to_csv(path_or_buf=op_address('0'), sep='\\t', index=False, encoding='utf-8')\n",
    "# print('message! >>> '+op_address('0')+' added.')\n",
    "\n",
    "\n",
    "# version 1.0 : 다음 column 들을 필터링.\n",
    "filt = ['First amino acid', 'Second amino acid', 'Second last amino acid', 'Last amino acid',\n",
    "        'A Count', 'R Count', 'N Count', 'D Count', 'C Count', 'Q Count', 'E Count', 'G Count', 'H Count', 'I Count', 'L Count',\n",
    "        'K Count', 'M Count', 'F Count', 'P Count', 'S Count', 'T Count', 'W Count', 'Y Count', 'V Count', 'U Count', 'O Count',\n",
    "        'Length', 'Oxidation (M) site IDs', 'Taxonomy IDs']\n",
    "\n",
    "df.drop(columns=filt, inplace=True)\n",
    "print('message! >>> 29 columns are dropped')\n",
    "\n",
    "v = op_address(1)\n",
    "df.to_csv(path_or_buf=v.pep_address2(), sep='\\t', index=False, encoding='utf-8')\n",
    "print('message! >>> '+v.pep_address2()+' added.')\n",
    "\n",
    "\n",
    "# index 넘버링 초기화\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c7070773-29df-4e04-9e59-7f06a0d610ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "82"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# amino acid 카운트 column drop. >>> find 메서드로 시퀀스에서 찾는 기능 도입하면 돼. df 크기 줄이는 목적.\n",
    "\n",
    "# Length drop: len(df[row, 'column_label'])메서드 사용.\n",
    "# ex) len(df.loc[0, 'Sequence'])\n",
    "\n",
    "# Fisrt amino acid / Second amino acid / Second last amino acid / Last amino acid 컬럼 drop\n",
    "# Sequence에서 앞뒤 첫,두번째 아미노산 의미.\n",
    "# ex) df.loc[0, 'Sequence'].count('A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf275f60-7a14-41f5-9e7b-9bc3d65f2e46",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "# filt = ['First amino acid', 'Second amino acid', 'Second last amino acid', 'Last amino acid',\n",
    "# 'A Count', 'R Count', 'N Count', 'D Count', 'C Count', 'Q Count', 'E Count', 'G Count', 'H Count', 'I Count', 'L Count',\n",
    "# 'K Count', 'M Count', 'F Count', 'P Count', 'S Count', 'T Count', 'W Count', 'Y Count', 'V Count', 'U Count', 'O Count', 'Length']\n",
    "\n",
    "# filt 요소 구현\n",
    "\n",
    "# (1-2) Nth order amino acid\n",
    "tmp = df.loc[0, 'Sequence']\n",
    "tmp[N-1]\n",
    "\n",
    "# (3-4) Nth order last amino acid\n",
    "tmp = df.loc[0, 'Sequence']\n",
    "tmp[-N]\n",
    "\n",
    "# (5-26) Count specific amino acid in sequence\n",
    "df.loc[0, 'Sequence'].count('A')\n",
    "\n",
    "# (27) Length of Sequence\n",
    "len(df.loc[0, 'Sequence'])\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10292400-c299-4aac-a0b1-475ba4524ed3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ae7fb1-41da-46f9-abc5-26239dabdc1b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
